src/Language/PureScript/CoreFn/Laziness.hs

```diff
@@ -0,0 +1,504 @@
+
+  -- And this is the function that prevents the search from actually being
+  -- infinite. It applies a filter to a `MaxRoseTree` at every level, looking for
+  -- indices anywhere in the tree that match the current vertex. If a match is
+  -- found with greater force than the current force, that part of the tree is
+  -- rewritten to abort; otherwise, that part of the tree is rewritten to be
+  -- empty (there's nothing new in that part of the search).
+  --
+  -- A new version of `cutLoops` is applied for each node in the search, so
+  -- each edge in a search path will add another filter on a new index. Since
+  -- there are a finite number of indices in our universe, this guarantees that
+  -- the analysis terminates, because no single search path can have length
+  -- greater than maxIdx.
+  cutLoops :: (Int, force) -> MaxRoseTree m force -> MaxRoseTree m force
+  cutLoops (i, force) = go
+    where
+    go = (=<<) . IM.traverseWithKey $ \i' (MaxRoseNode force' inner) ->
+      MaxRoseNode force' <$> if i == i' then guard (force >= force') $> pure IM.empty else pure $ go inner
+
+-- One last data structure to define and then it's on to the main event.
+--
+-- The laziness transform effectively takes a list of eager bindings (x = ...)
+-- and splits some of them into lazy definitions ($lazy_x = ...) and lazy
+-- bindings (x = $lazy_x ...). It's convenient to work with these three
+-- declarations as the following sum type:
+
+data RecursiveGroupItem e = EagerBinding Ann e | LazyDefinition e | LazyBinding Ann
+  deriving Functor
+
+-- |
+-- Transform a recursive binding group, reordering the bindings within when a
+-- correct initialization order can be statically determined, and rewriting
+-- bindings and references to be lazy otherwise.
+--
+applyLazinessTransform :: ModuleName -> [((Ann, Ident), Expr Ann)] -> ([((Ann, Ident), Expr Ann)], Any)
+applyLazinessTransform mn rawItems = let
+
+  -- Establish the mapping from names to ints.
+  rawItemsByName = M.fromList $ (snd . fst &&& first fst) <$> rawItems
+  maxIdx = M.size rawItemsByName - 1
+  rawItemsByIndex = A.listArray (0, maxIdx) $ M.elems rawItemsByName
+  names = M.keysSet rawItemsByName
+
+  -- Now do the first delay/force traversal of all the bindings to find
+  -- references to other names in this binding group.
+  --
+  -- The parts of this type mean:
+  -- D is the maximum force (or Nothing if unknown) with which the identifier C
+  -- is referenced in any delay-B position inside the expression A.
+  --
+  -- where A, B, C, and D are as below:
+  --                A           B (keys)           C (keys)           D
+  findReferences :: Expr Ann -> IM.MonoidalIntMap (IM.MonoidalIntMap (Ap Maybe (Max Int)))
+  findReferences = (getConst .) . onVarsWithDelayAndForce $ \delay force _ -> \case
+    Qualified mn' ident | all (== mn) mn', Just i <- ident `S.lookupIndex` names
+      -> Const . IM.singleton delay . IM.singleton i $ coerceForce force
+    _ -> Const IM.empty
```

Comment:
To clarify this with our current syntax of `#` (delay) and `!` (force), A, B, C, and D refer to this, right?
```
C = ...
A = C#B!D
```

Reply:
Oh wait... that's not quite right. `A` is an expression and part of it contains `C#B!D`.

Reply:
Yeah, I'm not sure how helpful the syntactical annotations are for this one, but an example would be
```
exprA = ... refC1#B1!D1 ... refC1#B1!D2 ... refC1#B2!D3 ... refC2#B1!D4 ... refC2#B1!* ...

findReferences <initializer of exprA> =
  IM.fromList [(B1, IM.fromList [(C1, Just (D1 `max` D2))
                                ,(C2, Nothing)
                                ])
              ,(B2, IM.fromList [(C1, Just D3)])
              ]
```